### DISCUSSION TITLE
Duda anexo de inyección de dependencias

### DISCUSSION THREAD

**Discussion by francorojo** [dds-jv/foro#133] [Consultas de objetos] [answered]
Buenaas, despues de releer el apunte me surgio una duda de la implementacion sobre la discusión del uso de una biblioteca de fechas.
Creo que por lo que pude entender, no seria correcto usar una inyeccion de dependencia en ese caso porque el objeto no es externo, por lo tanto no implica costos que no queremos al testearlo o por ejemplo retrasos de tiempo por demoras en como funciona el mismo.
Pero segun lo que entendi de el debate, capaz que mal, tampoco habria que usar metodos de clase estaticos para usar las funcionalidades de la biblioteca, ya que la interfaz no era expresiva.
Que opcion seria recomendable? Usar un adapter tal vez, para no tener una inyeccion de dependencia innecesaria y poder comunicarme de forma amigable con la biblioteca?

**Comment by francorojo** [comment]
Trate de comparar los ejemplo que paso Franco, y vi que el SendGrip seria externo, podria tenerse una inyeccion de dependencia, pero el JodaTime es una biblioteca que se podria importar y utilizar directamente sin preocuparnos porque es su idea, poder usar las veces que queramos los metodos que nos provee. Pero me llego una duda al ver el link:
![image](https://user-images.githubusercontent.com/49093389/83281119-fcbed380-a1ad-11ea-9d94-82fb3fa034df.png)
Me parece entender que esta usando metodos de clase, eso no era lo que no se podia usar directamente segun la discucion?

**Comment by gastonprieto** [comment]
> Creo que por lo que pude entender, no seria correcto usar una inyeccion de dependencia en ese caso porque el objeto no es externo, por lo tanto no implica costos que no queremos al testearlo o por ejemplo retrasos de tiempo por demoras en como funciona el mismo.

Acordate nada es absoluto primero. Lo que Franco trata de distinguir en el post es que hay unA diferencia entre los componentes externos, uno interactua con un sistema y el otro es una abstracción de una biblioteca que representa un tipo en particular. La naturaleza de estas es diferente.

Entonces es mas probable que quieras parametrizar el uso de algo que interactua con el sistema externo, porque te va a dejar hacer mejores pruebas unitarias. ¿Por qué?

> ... estas bibliotecas *dependen muchas veces* de singletons, que pueden mas o menos pesados de construir y luego los mantenemos vivos por toda la vida del sistema, y son por todo esto *grandes candidatos* a utilizar los patrones de inyección de dependencias o service locators explicados en ese apunte. 

----
Y de las fechas dice:
> ... no *suelen* tener traernos grandes problemas relacionados con la instanciación: creamos sus objetos libremente en cualquier parte del código, porque es fácil, barato e inocuo. 

Ahora.... que pasa si queres testear algo que depende del tiempo?? la única forma de que deje de ser un valor fijo es parametrizando al instanciador de fechas o parametrizando la fecha que utiliza esa operación, o sea caes también en la inyección de las dependencias.

Ejemplo: https://github.com/dds-utn/online-banking/blob/db35f3d21736a1a95f63f498d1cfac940c63e489/src/test/java/UsuarioTest.java#L37

-------
TLDR;
Básicamente la usas cuando vos quieras, incluso para inyectar objetos tuyos de dominio que tal vez no tenes implementados aún o porque no queres acoplarte a ese comportamiento real.

**Comment by gastonprieto** [comment]
> Que opcion seria recomendable? Usar un adapter tal vez, para no tener una inyeccion de dependencia innecesaria y poder comunicarme de forma amigable con la biblioteca?

Esto no creo que este muy relacionado, la inyección de dependencias es algo que usas para pametrizar las dependencias en la construcción.

Y hay un adapter cuando el objetivo de tu componente es adaptar dos interfaces distintas o cuando queres diseñar un componente que sabes que va a tener interfaces distintas pero aún no las conoces. 

O sea no son excluyentes entre sí, son cosas conceptos que atacan problemas distintos.

**Comment by gastonprieto** [comment]
Ese código de ejemplo de donde sale?

---- En general sobre los issues.
Danos un toque mas de contexto del apunte o el link que estas leyendo (yo sé a cual te referís), pero dentro de 1 año cuando alguien venga a leer el issue le va a faltar esa información.

Si bien dudo que alguien vaya a buscar por el código que pusiste, en el issue, es buena práctica cuando pongas código usar el markdown y subir la foto, porque así alguien mas puede encontrar si busca por ejemplo con la palabra "Joda"

**Comment by francorojo** [comment]
Aaa bueno lo agrego por las dudas https://www.codesolt.com/tutoriales/bibliotecas/joda-time/
Entonces puede ser que la mejor respuesta es depende, no? 
En el caso de un sistema externo, es probable por cuestiones de costos y tiempos, que se reciba como dependencia.
En el caso de una biblioteca, los metodos ofrecidos son mas faciles y "baratos" en costos, tal vez se use directamente a traves de los metodos de clase que provee, por ejemplo DataWithJoda, o tal vez a veces se instancie un objeto y en alguna parte termine siendo una inyeccion de dependencias de todas formas.
Lo que me queda es una duda para cerrar la idea, que tal vez no la comprendo yo porque no tengo el contexto de ese momento, perdon, aviseme si es asi por favor, o no tal vez sigo sin poder cerrar la idea. En el momento donde dice: "Usar la biblioteca de fechas como si fuera una clase con métodos estáticos", eso es malo? capaz depende el contexto se podria usar una opcion mejor? porque vi la implementacion de DataWithJoda y me parecia que hacia eso en el link, tomaba las funcionalidades que ofrece la biblioteca a traves de los metodos de clase o instanciando un objeto.
(Adjunto el link de donde tomo la fuente del issue: https://groups.google.com/forum/#!topic/dds-jv-cursada/iBL2dASE6lM)

**Comment by gastonprieto** [comment]
> En el caso de una biblioteca, los metodos ofrecidos son mas faciles y "baratos" en costos, tal vez se use directamente a traves de los metodos de clase que provee, por ejemplo DataWithJoda

Acá quiero aclarar algo `DateWithJoda` no es algo de Joda Time, es una clase que hizo el que escribio el post, es como si vos te hubieras hecho una clase `SuperManejadorDeFechas` y le hubieras puesto esos mensajes. 

Haciendo un poco analisis de esa clase veo que es poco cohesiva, fijate que tiene funcionalidades mixtas.
(de visualización) sabe como formatear un día de la semana. 
(creacionales) Sabe como adicionar meses a un día. Sabe como crear días nuevos en base a esos 3 parametros, que en realidad ya existe eso en Java8

Y justo cuando inicia la conversación original del post del grupo google, sobre los errores de elegir la versión de métodos de clase, era precisamente hacia esto donde apuntabamos que esta mal elegida. En vez de usar la API que da Joda o las fechas de Java, que es una interfaz de objetos, preferían usar una API mas procedural como la que estamos viendo en ese post que pasas de ejemplo.  

¿Ahora se entiende un poco mas a lo que iba al post?

**Comment by francorojo** [comment, marked as answer]
Perfectooo, muchas gracias por el tiempo y la explicacion!!!
